certora/specs/Splits.spec

```diff
@@ -0,0 +1,418 @@
+//// 
+//// ### Properties
+
+
+/// @notice Calculate results of splitting an amount using the current splits configuration.
+/// We verify that amount == collectableAmt + splitAmt, including that no overflow occurs
+rule correctnessOfSplitResults() {
+    env e; uint256 userId; uint128 amount; uint128 collectableAmt; uint128 splitAmt;
+
+    collectableAmt, splitAmt = splitResults(e, userId, true, amount);
+
+    assert amount >= collectableAmt;
+    assert amount >= splitAmt;
+    assert amount == collectableAmt + splitAmt;
+}
+
+
+/// @notice Splits user's received but not split yet funds among receivers.
+/// @param collectableAmt The amount made collectable for the user
+/// @param splitAmt The amount split to the user's splits receivers
+rule correctnessOfSplit() {
+    env e; uint256 userId; uint256 assetId; uint128 collectableAmt; uint128 splitAmt;
+
+    uint128 splittableBefore;   uint128 collectableBefore;
+    uint128 splittableAfter;    uint128 collectableAfter;
+    
+    splittableBefore = splittable(e, userId, assetId);
+    collectableBefore = collectable(e, userId, assetId);
+
+    collectableAmt, splitAmt = split(e, userId, assetId, true);
+
+    splittableAfter = splittable(e, userId, assetId);
+    collectableAfter = collectable(e, userId, assetId);
+
+    assert splittableBefore >= splittableAfter;
+    assert collectableBefore <= collectableAfter;
+    assert splittableBefore + collectableBefore >= splittableAfter + collectableAfter;
+}
+
+
+/// @notice Collects user's received already split funds
+/// After collection, no more collectable balance should be immediately available
+rule integrityOfCollect() {
+    env e; uint256 userId; uint256 assetId;
+    uint128 collectedAmt; uint128 collectableBefore; uint128 collectableAfter;
+
+    collectableBefore = collectable(e, userId, assetId);
+    collectedAmt = collect(e, userId, assetId);
+    collectableAfter = collectable(e, userId, assetId);
+
+    assert collectableAfter == 0;
+}
+
+
+/// @notice Calling the method collect() should never revert
+rule revertCharacteristicsOfCollect() {
+    env e; uint256 userId; uint256 assetId; uint128 collectedAmt;
+
+    require e.msg.sender != 0;  // safe assumption that prevents revert
+    require e.msg.value == 0;  // safe assumption that prevents revert
+
+    collectedAmt = collect@withrevert(e, userId, assetId);
+    assert !lastReverted;
+}
+
+
+/// @notice The method give() gives funds from the user to the receiver.
+/// @param amt The given amount
+/// After give of amt, the splittable of the receiver should increase exactly by amt
+rule correctnessOfGive() {
+    env e; uint256 userId; uint256 receiver; uint256 assetId; uint128 amt;
+    uint128 splittableBefore; uint128 splittableAfter;
+
+    splittableBefore = splittable(e, receiver, assetId);
+    give(e, userId, receiver, assetId, amt);
+    splittableAfter = splittable(e, receiver, assetId);
```

Comment:
What happens if you use a different environment variable for the second call to splittable? is this violated?

Reply:
The rule passes.
If there was an active dripper and we were testing the whole system, probably it should have not passed. In our case it probably shouldn't matter, but in general I think that the same e should be used.
Is it wrong to use the same `e` here?